### DISCUSSION TITLE
Seeking help on Implementing Surround Support in Vim Mode

### DISCUSSION THREAD

**Discussion by Kalmaegi** [zed-industries/zed#8018] [Feature Requests]
Hi everyone,

I'm looking to add surround support for Zed in Vim mode. After examining the Vim mode-related code, I noticed that currently, we match operations based on each input. For example, when we input ciw, we push a change onto the stack, then push an operator representing inner operation, and finally, when we input w, we match it as a Word action. If this action were in normal mode, it would pop the change and inner from the stack, then call the change method.

This is a simple example. Now, I realize that to achieve nvim's surround behavior, I would need to input ysiw". I've set ys as a custom surround method, and i remains the same. However, when executing w, if I want the Vim state to remain in a suspended state for receiving the next command, I can only check if the ys command exists at the top of the stack. Then, I combine the top i command with the current w command into one command. When the next input " is received, I pop the ys and iw commands from the stack, passing iw and " parameters to the surround method. This partially modifies the logic for the current input w action (and potentially other input actions). Additionally, this modification requires inputting all supported surround symbols into vim.json to set the methods called by input. I find this modification approach not very elegant, so I'm seeking help to confirm two things:

Do I have to set trigger methods for each surround symbol individually in vim.json if I want to support surround functionality? If there's a convenient method like nvim's lhs and rhs, I could configure them as the content for rhs input, which would be more flexible.

Taking ysiw" as an example, ys is convenient as one operation, but i, w, " are separate operations. Currently, do I have to treat them as four separate operations and determine their behavior, especially during w operation, based on whether ys is active? This might not be conducive to maintenance, but I haven't thought of a more convenient method.

Thank you all for your help; it's greatly appreciated.

this input 'ciw' called function place:
https://github.com/zed-industries/zed/blob/ca251babcd408d4a6944652e524bfed45cfc6a9e/crates/vim/src/object.rs#L56-L61
https://github.com/zed-industries/zed/blob/ca251babcd408d4a6944652e524bfed45cfc6a9e/crates/vim/src/object.rs#L87-L95
https://github.com/zed-industries/zed/blob/ca251babcd408d4a6944652e524bfed45cfc6a9e/crates/vim/src/normal.rs#L160-L165

https://github.com/zed-industries/zed/blob/ca251babcd408d4a6944652e524bfed45cfc6a9e/crates/vim/src/normal/change.rs#L17

https://github.com/zed-industries/zed/blob/ca251babcd408d4a6944652e524bfed45cfc6a9e/crates/vim/src/normal/change.rs#L60

**Comment by Kalmaegi** [comment]
I have a tentative idea for these two points, because in most cases the surround is valid for the current word, so for example, `ys i w "`can be seen as `ys i "`, so that there is no need to change the current call stack, and I found that there are shortcut settings for many parentheses in the vim.json, I can reuse these, and I will submit a PR as soon as possible

**Comment by JosephTLyons** [comment]
I'm a bit late here, but looks like you've implemented this now:

- https://github.com/zed-industries/zed/pull/9400

Let's close this Discussion out!
